gauge: add plugin support

NyCodeGHG opened the pull request:
## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

Adds nix managed plugin support to gauge.

Allows to get a gauge with installed plugins using `gauge.withPlugins (p: with p; [ js html-report ])`
Also allows reading plugins from a gauge manifest.json file: `gauge.fromManifest ./manifest.json`.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

NyCodeGHG committed:
gaugePlugins.makeGaugePlugin: init

NyCodeGHG committed:
gaugePlugins.java: init at 0.10.3

NyCodeGHG committed:
gauge: add wrapper

Adds a wrapper to the gauge package, which allows installing plugins declaratively with nix.

NyCodeGHG committed:
gaugePlugins.html-report: init at 4.3.1

NyCodeGHG committed:
gauge: add marie to maintainers

NyCodeGHG committed:
gaugePlugins.dotnet: init at 0.5.7

NyCodeGHG committed:
gauge: add passthru.withPlugins helper function

NyCodeGHG committed:
gaugePlugins.ruby: init at 0.8.0

NyCodeGHG committed:
gaugePlugins.js: init at 4.0.1

NyCodeGHG committed:
gaugePlugins.go: init at 0.4.0

NyCodeGHG committed:
gaugePlugins.screenshot: init at 0.2.0

NyCodeGHG committed:
gauge: add fromManifest helper function

NyCodeGHG committed:
gaugePlugins.xml-report: init at 0.5.1

NyCodeGHG commented:
```
- Attribute `pkgs.gauge-unwrapped` is a new top-level package using `pkgs.callPackage ./pkgs/development/tools/gauge { /* ... */ }`.
  Please define it in pkgs/by-name/ga/gauge-unwrapped/package.nix instead.
  See `pkgs/by-name/README.md` for more details.
  Since the second `callPackage` argument is `{ }`, no manual `callPackage` in pkgs/top-level/all-packages.nix is needed anymore.

This PR introduces additional instances of discouraged patterns as listed above. Merging is discouraged but would not break the base branch.
```

hmm, I would rather keep all the gauge stuff at the same place tbh

NyCodeGHG commented:
@vdemeester could you review this as the existing maintainer of gauge?
That would be awesome! :)

vdemeester reviewed (approved):
LGTM from me 👼🏼

NyCodeGHG commented:
removed the add mainProgram commit, as this has been already added in master

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/3768

NyCodeGHG commented:
uncommented code which was still commented out from testing
html-report updated to 4.3.1
screenshot updated to 0.2.0
xml-report updated to 0.5.1
removed windows platforms from plugins
rebased onto master

AndersonTorres commented:
> hmm, I would rather keep all the gauge stuff at the same place tbh

can't it be done by using `passthru`?

SuperSandro2000 commented on pkgs/development/tools/gauge/default.nix:
```diff
@@ -17,9 +23,8 @@ buildGoModule rec {
 
   meta = with lib; {
     description = "Light weight cross-platform test automation";
-    mainProgram = "gauge";
```
Why should we drop this?

NyCodeGHG replied:
we shouldn't, i messed something up while rebasing

SuperSandro2000 commented on pkgs/development/tools/gauge/plugins/default.nix:
```diff
@@ -0,0 +1,12 @@
+{ lib, pkgs }:
+lib.makeScope pkgs.newScope (final: {
+  makeGaugePlugin = final.callPackage ./make-gauge-plugin.nix { };
+  dotnet = final.callPackage ./dotnet { };
```
Please use with or inherit to avoid the repeated finals.

AndersonTorres replied:
Imho there is no problem being explicit here.

however `with` is bad.

SuperSandro2000 commented on pkgs/development/tools/gauge/plugins/make-gauge-plugin.nix:
```diff
@@ -0,0 +1,104 @@
+  inherit (if isMultiArch then data.${system} else data) url hash;
+in
+stdenvNoCC.mkDerivation (finalAttrs: (lib.recursiveUpdate {
+  pname = "gauge-plugin-${name}";
```
```suggestion
  pname = "gauge-plugin-${pname}";
```
Maybe less confusing

SuperSandro2000 commented on pkgs/development/tools/gauge/wrapper.nix:
```diff
@@ -0,0 +1,56 @@
+
+  installPhase = ''
+    mkdir -p $out{bin,/share/gauge/{plugins,config}}
+    export GAUGE_HOME="$out/share/gauge"
```
If we need to remove logs afterwards, maybe it is easier to only copy over, what we do need?

NyCodeGHG replied:
done!

SuperSandro2000 commented on pkgs/development/tools/gauge/plugins/xml-report/default.nix:
```diff
@@ -0,0 +1,25 @@
+
+  repo = "getgauge/xml-report";
+  releasePrefix = "xml-report-";
+  systemMap = {
```
It seems like this map is very similar for every plugin except that not all support all platforms. Maybe the presence of an a download url could be enough to map this automatically?

NyCodeGHG replied:
done!

AndersonTorres commented:
I have wrote something similar in the `windowmaker.passthru.dockapps`.
Maybe it can be replicated.

NyCodeGHG commented:
@AndersonTorres 
> I have wrote something similar in the `windowmaker.passthru.dockapps`. Maybe it can be replicated.

hm, I'm not sure.
Is there some new convention to do this instead?
The plugins don't really have a direct relationship to which gauge package is used.
Also this breaks search.nixos.org, no?

SuperSandro2000 commented on pkgs/development/tools/gauge/plugins/make-gauge-plugin.nix:
```diff
@@ -0,0 +1,94 @@
+
+    baseUrl="https://github.com/${repo}/releases/download/$latestTag/${releasePrefix}$latestVersion"
+
+    ${if !isCrossArch then
+        lib.concatStringsSep "\n" (map (platform: ''updateSystem "${platform}" "''${baseUrl}-${systemMap.${platform}}.zip"'') meta.platforms)
+      else
+        "updateSingle \${baseUrl}.zip"
+    }
```
```suggestion
    ${if isCrossArch then
        "updateSingle \${baseUrl}.zip"
      else
        lib.concatStringsSep "\n" (map (platform: ''updateSystem "${platform}" "''${baseUrl}-${systemMap.${platform}}.zip"'') meta.platforms)
    }
```

yayayayaka commented:
Result of `nixpkgs-review pr 295283` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>10 packages built:</summary>
  <ul>
    <li>gauge</li>
    <li>gauge-unwrapped</li>
    <li>gaugePlugins.dotnet</li>
    <li>gaugePlugins.go</li>
    <li>gaugePlugins.html-report</li>
    <li>gaugePlugins.java</li>
    <li>gaugePlugins.js</li>
    <li>gaugePlugins.ruby</li>
    <li>gaugePlugins.screenshot</li>
    <li>gaugePlugins.xml-report</li>
  </ul>
</details>

yayayayaka reviewed (approved):
Reviewed points:
- diff looking good to me
- release note entry also looking good to me
- successfully built a custom `gauge.withPlugins` as hinted in the PR description
- nixpkgs-review also happy
- approved by existing gauge maintainer.

In my opinion, this looks ready to be merged and I would be willing to hit the merge button soon.

NyCodeGHG commented:
rebased onto master to resolve merge conflicts
